src/hotspot/share/gc/g1/g1YoungGCPreEvacuateTasks.cpp

```diff
@@ -0,0 +1,192 @@
+    _local_refinement_stats = NEW_C_HEAP_ARRAY(G1ConcurrentRefineStats, _num_workers, mtGC);
+
+    for (uint i = 0; i < _num_workers; i++) {
+      _local_tlab_stats[i].reset();
+      _local_refinement_stats[i].reset();
+    }
+  }
+
+  ThreadLocalAllocStats tlab_stats() const {
+    ThreadLocalAllocStats result;
+    for (uint i = 0; i < _num_workers; i++) {
+      result.update(_local_tlab_stats[i]);
+    }
+    return result;
+  }
+
+  G1ConcurrentRefineStats refinement_stats() const {
+    G1ConcurrentRefineStats result;
+        for (uint i = 0; i < _num_workers; i++) {
+      result += _local_refinement_stats[i];
+    }
+    return result;
+  }
+};
+
+class G1PreEvacuateCollectionSetBatchTask::NonJavaThreadFlushLogs : public G1AbstractSubTask {
+  RetireTLABAndFlushLogsClosure _tc;
+
+public:
+  NonJavaThreadFlushLogs() : G1AbstractSubTask(G1GCPhaseTimes::NonJavaThreadFlushLogs), _tc() { }
+
+  void do_work(uint worker_id) override {
+    Threads::non_java_threads_do(&_tc);
+  }
+
+  double worker_cost() const override {
+    return 1.0;
+  }
+
+  ThreadLocalAllocStats tlab_stats() const { return _tc._tlab_stats; }
+  G1ConcurrentRefineStats refinement_stats() const { return _tc._refinement_stats; }
+};
+
+G1PreEvacuateCollectionSetBatchTask::G1PreEvacuateCollectionSetBatchTask() :
+  G1BatchedTask("Pre Evacuate Prepare", G1CollectedHeap::heap()->phase_times()),
+  _old_pending_cards(G1BarrierSet::dirty_card_queue_set().num_cards()),
+  _java_stats(new JavaThreadRetireTLABAndFlushLogs()),
+  _non_java_stats(new NonJavaThreadFlushLogs()) {
+
+  // Disable mutator refinement until concurrent refinement decides otherwise.
+  G1BarrierSet::dirty_card_queue_set().set_mutator_refinement_threshold(SIZE_MAX);
+
+  add_serial_task(_non_java_stats);
+  add_parallel_task(_java_stats);
+}
+
+static void verify_empty_dirty_card_logs() {
+#ifdef ASSERT
+  ResourceMark rm;
+
+  struct Verifier : public ThreadClosure {
+    size_t _buffer_size;
+    Verifier() : _buffer_size(G1BarrierSet::dirty_card_queue_set().buffer_size()) {}
+    void do_thread(Thread* t) override {
+      G1DirtyCardQueue& queue = G1ThreadLocalData::dirty_card_queue(t);
+      assert((queue.buffer() == nullptr) || (queue.index() == _buffer_size),
+             "non-empty dirty card queue for thread %s", t->name());
+    }
+  } verifier;
+  Threads::threads_do(&verifier);
+#endif
+}
+
+G1PreEvacuateCollectionSetBatchTask::~G1PreEvacuateCollectionSetBatchTask() {
+  ThreadLocalAllocStats total_tlab_stats;
+  total_tlab_stats.update(_java_stats->tlab_stats());
+  total_tlab_stats.update(_non_java_stats->tlab_stats());
+  total_tlab_stats.publish();
+
+  G1DirtyCardQueueSet& qset = G1BarrierSet::dirty_card_queue_set();
+
+  G1ConcurrentRefineStats total_refinement_stats;
+  total_refinement_stats += _java_stats->refinement_stats();
+  total_refinement_stats += _non_java_stats->refinement_stats();
+  qset.update_refinement_stats(total_refinement_stats);
+
```

Comment:
Both are actually the tasks that after adding them to the batchtask are owned and deleted by it. I'll rename the variables because otherwise it is confusing (I went back and forth with the names of these multiple times and apparently chose the more confusing option).